UserDetailsClient/UserDetailsClient.Core/Features/LogOn/B2CAuthenticationService.cs

```diff
@@ -0,0 +1,157 @@
+                .WithPrompt(Prompt.NoPrompt)
+                .WithAuthority(B2CConstants.AuthorityPasswordReset)
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> EditProfile()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            // KNOWN ISSUE:
+            // User will get prompted 
+            // to pick an IdP again.
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithAccount(GetAccountByPolicy(accounts, B2CConstants.PolicyEditProfile))
+                .WithPrompt(Prompt.NoPrompt)
+                .WithAuthority(B2CConstants.AuthorityEditProfile)
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> SignIn()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithAccount(GetAccountByPolicy(accounts, B2CConstants.PolicySignUpSignIn))
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var newContext = UpdateUserInfo(ar);
+            return newContext;
+        }
+
+        public async Task<UserContext> SignOut()
+        {
+
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            while (accounts.Any())
+            {
+                await PCA.RemoveAsync(accounts.FirstOrDefault());
+                accounts = await PCA.GetAccountsAsync();
+            }
+            var signedOutContext = new UserContext();
+            signedOutContext.IsLoggedOn = false;
+            return signedOutContext;
+        }
+
+        private IAccount GetAccountByPolicy(IEnumerable<IAccount> accounts, string policy)
+        {
+            foreach (var account in accounts)
+            {
+                string userIdentifier = account.HomeAccountId.ObjectId.Split('.')[0];
+                if (userIdentifier.EndsWith(policy.ToLower())) return account;
+            }
+
+            return null;
+        }
+
+
+        private string Base64UrlDecode(string s)
+        {
+            s = s.Replace('-', '+').Replace('_', '/');
+            s = s.PadRight(s.Length + (4 - s.Length % 4) % 4, '=');
+            var byteArray = Convert.FromBase64String(s);
+            var decoded = Encoding.UTF8.GetString(byteArray, 0, byteArray.Count());
+            return decoded;
+        }
+
+        public UserContext UpdateUserInfo(AuthenticationResult ar)
```

Comment:
This method might be better suited on the UserContext class.

Reply:
prefer to keep mapping logic separate from entity (User Context object)

Reply:
done